src/atomsbase_interface.jl

```diff
@@ -0,0 +1,58 @@
+#
+# Interface between AtomsBase.jl and GeometryOptimization.jl that provides 
+# utility functions for manipulating systems.
+#
+# IMPORTANT: Note that we always work in cartesian coordinates.
+#
+export update_positions, update_not_clamped_positions, clamp_atoms
+
+
+@doc raw"""
+Creates a new system based on ``system`` but with atoms positions updated 
+to the ones provided.
+
+"""
+function update_positions(system, positions::AbstractVector{<:AbstractVector{<:Unitful.Length}})
+    particles = [Atom(atom; position) for (atom, position) in zip(system, positions)]
+    AbstractSystem(system; particles)
+end
+
+@doc raw"""
+Creates a new system based on ``system`` where the non clamped positions are
+updated to the ones provided (in the order in which they appear in the system).
+
+"""
+function update_not_clamped_positions(system, positions::AbstractVector{<:Unitful.Length})
+    mask = not_clamped_mask(system)
+    new_positions = deepcopy(position(system))
+    new_positions[mask] = reinterpret(reshape, SVector{3, eltype(positions)},
+                                      reshape(positions, 3, :))
+    update_positions(system, new_positions)
+end
+
+@doc raw"""
+Returns a mask for selecting the not clamped atoms in the system.
+
+"""
+function not_clamped_mask(system)
+    # If flag not set, the atom is considered not clamped.
+    [haskey(a, :clamped) ? !a[:clamped] : true for a in system]
+end
+
+function not_clamped_positions(system)
+    mask = not_clamped_mask(system)
+    Iterators.flatten(system[mask, :position])
+end
+
+@doc raw"""
+    Clamp given atoms in the system. Clamped atoms are fixed and their positions 
+    will not be optimized. The atoms to be clamped should be given as a list of 
+    indices corresponding to their positions in the system.
+
+    """
+function clamp_atoms(system, clamped_indexes::Union{AbstractVector{<:Integer},Nothing})
+    clamped = falses(length(system))
+    clamped[clamped_indexes] .= true
+    particles = [Atom(atom; clamped=m) for (atom, m) in zip(system, clamped)]
```

Comment:
An alternative way to do this (see also @mfherbst's [related comment](https://github.com/JuliaMolSim/GeometryOptimization.jl/pull/1/files#r1431521080)) that would be a bit more like what Optim.jl or Optimization.jl do would be to pass a separate `constraints` argument to the optimizer. I lean towards this approach because it allows for constraints that might not be a property of a single atom the way "being clamped" is – e.g. bond or symmetry constraints.

Reply:
Just now noticed https://github.com/JuliaMolSim/GeometryOptimization.jl/issues/2 – I think this would also substantially simplify addressing that...

Reply:
Hi Rachel. Thanks for the feedback. 
For the constraints, I suggest we defer implementating more fancy one to #2 and just work with fixed/clamped at the moment. 

Also, in terms of speed and dimensionality, I suspect that the approach we now have with `clamped` is useful, since it explicitly reduces the dimension of the problem before starting any optimization, whereas implementing this with a constraint in Optimization.jl will likely add some overhead. 

Long story short: I would advocate for keeping the clamping system for now, since it allows for efficient treatment of the basic situations. Then, once we have implemented #2 we can see if a separate function for the case of atoms clamping is still needed or not.

Reply:
Hmm, I see what you're saying, and I definitely appreciate the imperative to start simple and only build in complications if they're needed, but given that we know that things like bond or symmetry constraints are real applications, I guess I just want to avoid putting something in that we'd then need to roll back, or creating confusing redundancies. I'm okay proceeding with the way it is now if others are okay with this, particularly @mfherbst as the DFTK person...

Reply:
I would support keeping "clamps" as a special case. Even after we have more general constraints. In any case we want to make progress and this is a good start.

Reply:
I agree. The current state is a good compromise for a first solution.

Reply:
Alright, given comments above, I'm happy for this to be merged as-is. I merely reserve the right for an "I told you so" if we decide to change it later... 🤪